vpr/src/place/analytic_placer.cpp

```diff
@@ -0,0 +1,852 @@
+    EquationSystem(size_t rows, size_t cols) {
+        A.resize(cols);
+        rhs.resize(rows);
+    }
+
+    // A[col] is an entire column of the sparse matrix
+    // each entry in A[col][index] is a pair with {row_number, matrix value}.
+    //
+    // The strategy of skipping 0 row entries in each column enables easy conversion to
+    // Compressed Column Storage scheme supported by Eigen to reduce memory consumption
+    // and increase performance
+    std::vector<std::vector<std::pair<int, T>>> A;
+    // right hand side vector, i.e. b in Ax = b
+    std::vector<T> rhs;
+
+    // System of equation is reset by:
+    // Clearing all entries in A's column, but size of A (number of columns) is preserved
+    // right hand side vector is set to default value of its templated type
+    void reset() {
+        for (auto& col : A)
+            col.clear();
+        std::fill(rhs.begin(), rhs.end(), T());
+    }
+
+    // Add val to the matrix entry at (row, col)
+    // create entry if it doesn't exist
+    void add_coeff(int row, int col, T val) {
+        auto& A_col = A.at(col);
+        // Binary search for the row entry in column col
+        int begin_i = 0, end_i = int(A_col.size()) - 1;
+        while (begin_i <= end_i) {
+            int i = (begin_i + end_i) / 2;
+            if (A_col.at(i).first == row) {
+                A_col.at(i).second += val;
+                return;
+            }
+            if (A_col.at(i).first > row)
+                end_i = i - 1;
+            else
+                begin_i = i + 1;
+        }
+        A_col.insert(A_col.begin() + begin_i, std::make_pair(row, val));
+    }
+
+    // Add val to the "row"-th entry of right hand side vector
+    void add_rhs(int row, T val) { rhs[row] += val; }
+
+    // Solving Ax = b, using current x as an initial guess, returns x by reference.
+    // (x must be of correct size, A and rhs must have their entries filled in)
+    // tolerance is residual error from solver: |Ax-b|/|b|, 1e-5 works well,
+    // can be tuned in ap_cfg in AnalyticPlacer constructor
+    //
+    void solve(std::vector<T>& x, float tolerance) {
+        using namespace Eigen;
+
+        VTR_ASSERT(x.size() == A.size());
+
+        // Converting A into SparseMatrix format from Eigen
+        VectorXd vec_x_guess(x.size()), vec_rhs(rhs.size());
+        SparseMatrix<T> mat(A.size(), A.size());
+
+        std::vector<int> colnnz;
+        for (auto& A_col : A)
+            colnnz.push_back(int(A_col.size()));
+        mat.reserve(colnnz); // reserve memory for mat depending on number of entries in each row
+        for (int col = 0; col < int(A.size()); col++) {
+            auto& A_col = A.at(col);
+            for (auto& row_entry : A_col)
+                mat.insert(row_entry.first, col) = row_entry.second;
+        }
+
+        // use current value of x as guess for iterative solver
+        // feeding next solution back to solver and repeat can achieve better results
```

Comment:
sorry, feeding solution back to solver happends outside EquationSystem::solve in AnalyticPlacer::build_solve_direction. The exact number of build_solve_iter can be tuned in AP cfg. I will remove line 93 here to avoid confusion.